src/main/java/org/magicdgs/readtools/utils/distmap/DistmapGATKWriter.java

```diff
@@ -0,0 +1,146 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package org.magicdgs.readtools.utils.distmap;
+
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.GATKReadWriter;
+import scala.Tuple2;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.function.Consumer;
+
+/**
+ * Abstract class for GATKRead writers.
+ *
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class DistmapGATKWriter implements GATKReadWriter {
+
+    // the name of the source for the output
+    private final String sourceName;
+    // the writer stream
+    private final PrintStream writer;
+    // if it is single-end, this will just write the read out
+    // if it is pair-end, this will use DistmapPairedConsumer
+    private final Consumer<GATKRead> singleEndHandler;
+
+    /**
+     * Private constructor with a OutputStremWriter.
+     *
+     * @param writer     output stream to write the reads on.
+     * @param paired     if {@code true}, the file will be written as paired; otherwise,
+     *                   it will be written as single.
+     * @param sourceName the name for the source where we are writing to.
+     */
+    public DistmapGATKWriter(final PrintStream writer, final String sourceName,
+            final boolean paired) {
+        this.writer = writer;
+        this.sourceName = sourceName;
+        this.singleEndHandler = (paired)
+                ? new DistmapPairedConsumer(this::addPair)
+                : read -> printAndCheckError(DistmapEncoder.encode(read));
+    }
+
+    /**
+     * Adds a read to the writer. If added in pair-end mode, the reads in the pair are expected to
+     * be passed one after the other.
+     */
+    @Override
+    public void addRead(final GATKRead read) {
+        singleEndHandler.accept(read);
+    }
+
+    /**
+     * Writes a pair of reads in the Distmap format.
+     *
+     * @param pair the pair of reads.
+     */
+    public void addPair(final Tuple2<GATKRead, GATKRead> pair) {
+        printAndCheckError(DistmapEncoder.encode(pair));
+    }
+
+    /**
+     * Close the underlying writer.
+     *
+     * @throws IllegalStateException if the writer is in paired and not second read is added.
+     */
+    @Override
+    public void close() throws IOException {
+        // close the handler if it is a DistmapPairedConsumer
+        if (singleEndHandler instanceof Closeable) {
+            ((Closeable) singleEndHandler).close();
+        }
+        writer.close();
+    }
+
+    // helper method to print and with the writer and check if an error occurs to throw an exception
+    private void printAndCheckError(final String toPrint) {
+        writer.println(toPrint);
+        if (writer.checkError()) {
```

Comment:
Because this is going to be used mainly in a Hadoop clustr, I think that we can use instead a better approach to capture the real `IOException` if it occurs and debug the cause, such as network connection and so on.